[Benchmark] DiT Model Benchmark under Mixed Workloads

asukaqaq-s opened the pull request:
<!-- markdownlint-disable -->
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

This PR enhances the diffusion serving benchmarks with trace-driven mixed workloads and SLO evaluation.

Mixed workload support: the benchmark can replay heterogeneous requests in a single run, including different resolutions (width/height), number of frames, and inference steps, so it better reflects real-world serving traffic.
SLO metrics: the benchmark reports SLO attainment in addition to latency/throughput, enabling SLO-aware evaluation and regression tracking across models and deployments.
Dataset source: the trace/dataset used to generate these mixed requests is based on [https://arxiv.org/abs/2510.02838]

## Test Plan


We first perform offline execution to profile the model latency on the target GPU, and define the SLO as three times the measured execution time.
```
python -m benchmarks.text_to_image.offline_profile --dataset-path  benchmarks/text_to_image/dataset/sd3_trace.txt --output benchmarks/text_to_image/dataset/slo_define.json

```

Run server 

```
# run on single A100 80GB
vllm serve Qwen/Qwen-Image --omni --port 8099
```


Run benchmark client
```

## unchanged from before
python3 benchmarks/diffusion_benchmark_serving.py --base-url http://localhost:8099 \
    --model qwen/qwen-image --dataset trace --num-prompts 5 --request-rate 0.1 

## enable slo
python3 benchmarks/diffusion_benchmark_serving.py --base-url http://localhost:8099 \
    --model qwen/qwen-image --dataset trace --num-prompts 5 --request-rate 0.1 --slo 
 
## request rate 1
python3 benchmarks/diffusion_benchmark_serving.py --base-url http://localhost:8099 \
    --model qwen/qwen-image --dataset trace --num-prompts 5 --request-rate 1 --slo 

```

## Test Result

<img width="1027" height="356" alt="image" src="https://github.com/user-attachments/assets/a34609eb-80a8-4070-89b8-f369ebaed334" />
<img width="978" height="429" alt="image" src="https://github.com/user-attachments/assets/a1ad093c-f868-4d9f-80c6-3957755adcb6" />

<img width="839" height="432" alt="image" src="https://github.com/user-attachments/assets/6d5d1f24-65ef-4a27-8e9a-6cef1c4c236d" />

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

ZJY0516 reviewed (changes requested):
I don't want to upload trace to this repo. Could you please upload it to huggingface?

david6666666 commented:
Agree, @asukaqaq-s you  just need to modify the corresponding slo settings in benchmarks/diffusion_benchmark_serving.py, and delete everything else.

asukaqaq-s commented:
Thanks everyone! I plan to make the following updates and push them within the next couple of days:

1. Add SLO support to existing benchmarks
2. Introduce a traced dataset for benchmarking and upload it to Hugging Face
3. Improve configuration and SLO definition logic

hsliuustc0106 commented:
> Thanks everyone! I plan to make the following updates and push them within the next couple of days:
> 
> 1. Add SLO support to existing benchmarks
> 2. Introduce a traced dataset for benchmarking and upload it to Hugging Face
> 3. Improve configuration and SLO definition logic

hope this can be merged this week

asukaqaq-s commented:
Summary of Changes: Addressed the feedback from the last review:
1. Enabled downloading and parsing trace datasets from HuggingFace.
2. Updated width/height handling: uses dataset values if provided, otherwise falls back to defaults.
3. Added --slo and --slo-base-ms flags for SLO calculation and tracking.


Note on SLO Logic: Currently, slo_base_time denotes the SLO (ms) required for a single minimal inference unit. If left unspecified, it triggers an offline_profiling run. Please verify if this logic aligns with expectations.

ToDo:
[ ] Update tests and documentation.

asukaqaq-s commented:
> how do we test in a single run, including different resolutions (width/height), number of frames, and inference steps

@david6666666 

Sorry, I want to make sure I understand your question correctly.
Are you asking about:

- Testing multiple resolutions / frame counts / inference steps within a single request, or
- Running a single benchmark run over a dataset, where each request can have different resolutions, number of frames, and steps?
<img width="492" height="766" alt="image" src="https://github.com/user-attachments/assets/a81ec484-4d63-4c39-a31b-cd79874094c2" />

If you mean the latter, you can take a look at the trace dataset I uploaded to Hugging Face.
In the benchmark script, each request is parsed individually: during execution, the script first checks whether a request explicitly specifies parameters such as resolution, number of frames, or inference steps. If these fields are present, the per-request values from the dataset are used; otherwise, the benchmark falls back to the default configuration.

This allows a single benchmark run to naturally cover heterogeneous workloads with mixed resolutions, frame counts, and step settings.

david6666666 commented:
> > Yes, this dataset consists entirely of T2I requests. I also have separate T2V datasets in the same Hugging Face repository.
> 
> Sure — I’ll make the prompts different across requests, since this doesn’t affect performance. Varying the number of steps per request is less common in production, so I’ll keep steps fixed for now. Let me know if this approach looks reasonable to you.

Reasonable, If the dataset and pr is ready, tell me

asukaqaq-s commented on benchmarks/diffusion/diffusion_benchmark_serving.py:
Just to clarify, are you referring to the usage example shown at the top of diffusion_benchmark_serving.py, i.e.:

```
Usage:
    # Video
    t2v:
    python3 benchmarks/diffusion_benchmark_serving.py \
        --dataset vbench --task t2v --num-prompts 10 \
        --height 480 --width 640 --fps 16 --num-frames 80

```

david6666666 commented:
overall lgtm

ZJY0516 reviewed (approved):
just one little problem: do we need to put the readme in omni online doc? @hsliuustc0106
